### ISSUE TITLE
Mutex cannot be mocked since 3.9.0 (Stack level too deep)

### ISSUE THREAD

**Issue by sergio-bobillier** [rspec/rspec-mocks#1574]
### Mutex cannot be mocked since

Since `rspec 3.9.0` mocks of the `Mutex` class result in a `SystemStackError`: stack level too deep. Error. I unearthed these two issues:

* https://github.com/rspec/rspec-mocks/issues/1327
* https://github.com/rspec/rspec-support/pull/411

And I thought this had already been fixed and it was something related to my setup or the combination of gems in my project but actually the issue is still there. Reproducible with Ruby 2.x and 3.x

### Your environment
* Tested with: `3.2.2` and `2.7.7`
* rspec-mocks version: Tested with `3.9.0` all the way up to `3.13.0` the issue is visible in all of them.

### Steps to reproduce

I created this simple project to show the issue:

`writer.rb`:

```ruby
# frozen_string_literal: true

class Writer
  def initialize
    @mutex = Mutex.new
    @output = StringIO.new
  end

  def write(message)
    mutex.synchronize do
      output << message
    end
  end

  private

  attr_reader :mutex
end
```

`spec/writer_spec.rb`:

```ruby
# frozen_string_literal: true

require_relative '../writer'

RSpec.describe Writer do
  subject(:writer) { described_class.new }

  let(:mocked_mutex) do
    instance_double(
      Mutex
    )
  end

  before do
    allow(Mutex).to receive(:new).and_return(mocked_mutex)
    allow(mocked_mutex).to receive(:synchronize).and_yield
  end

  describe '#initialize' do
    it 'creates a new Mutex' do
      expect(Mutex).to receive(:new)
      writer
    end
  end
end
```

### Expected behavior

I would expect my test to pass (like it does with RSpec `3.8.0`)

```
.

Finished in 0.01108 seconds (files took 0.10239 seconds to load)
1 example, 0 failures
```

### Actual behavior

I'm getting the mentioned `SystemStackError: stack level too deep` with such a backtrace:

```
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-mocks-3.13.0/lib/rspec/mocks/method_reference.rb:117:in `method_implemented?'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-mocks-3.13.0/lib/rspec/mocks/method_reference.rb:48:in `block in defined?'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-mocks-3.13.0/lib/rspec/mocks/object_reference.rb:138:in `when_loaded'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-mocks-3.13.0/lib/rspec/mocks/method_reference.rb:47:in `defined?'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-mocks-3.13.0/lib/rspec/mocks/method_reference.rb:107:in `block in original_method'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-mocks-3.13.0/lib/rspec/mocks/object_reference.rb:138:in `when_loaded'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-mocks-3.13.0/lib/rspec/mocks/method_reference.rb:106:in `original_method'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-mocks-3.13.0/lib/rspec/mocks/method_reference.rb:53:in `with_signature'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-mocks-3.13.0/lib/rspec/mocks/verifying_proxy.rb:166:in `validate_arguments!'

    ... repeats meny hundred times ...

     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-mocks-3.13.0/lib/rspec/mocks/proxy.rb:199:in `message_received'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-mocks-3.13.0/lib/rspec/mocks/method_double.rb:98:in `proxy_method_invoked'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-mocks-3.13.0/lib/rspec/mocks/verifying_proxy.rb:161:in `proxy_method_invoked'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-mocks-3.13.0/lib/rspec/mocks/method_double.rb:74:in `block (2 levels) in define_proxy_method'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-mocks-3.13.0/lib/rspec/mocks/proxy.rb:164:in `reset'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-mocks-3.13.0/lib/rspec/mocks/proxy.rb:315:in `reset'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-mocks-3.13.0/lib/rspec/mocks/space.rb:79:in `block in reset_all'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-mocks-3.13.0/lib/rspec/mocks/space.rb:79:in `each_value'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-mocks-3.13.0/lib/rspec/mocks/space.rb:79:in `reset_all'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-mocks-3.13.0/lib/rspec/mocks.rb:52:in `teardown'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/mocking_adapters/rspec.rb:27:in `teardown_mocks_for_rspec'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:521:in `run_after_example'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:283:in `block in run'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:511:in `block in with_around_and_singleton_context_hooks'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:468:in `block in with_around_example_hooks'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:486:in `block in run'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:624:in `run_around_example_hooks_for'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:486:in `run'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:468:in `with_around_example_hooks'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:511:in `with_around_and_singleton_context_hooks'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:259:in `run'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:646:in `block in run_examples'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:642:in `map'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:642:in `run_examples'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:607:in `run'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:608:in `block in run'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:608:in `map'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:608:in `run'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:121:in `block (3 levels) in run_specs'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:121:in `map'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:121:in `block (2 levels) in run_specs'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/configuration.rb:2091:in `with_suite_hooks'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:116:in `block in run_specs'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/reporter.rb:74:in `report'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:115:in `run_specs'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:89:in `run'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:71:in `run'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:45:in `invoke'
     # /Users/sergio/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rspec-core-3.13.0/exe/rspec:4:in `<top (required)>'
     # /Users/sergio/.rbenv/versions/3.2.2/bin/rspec:25:in `load'
     # /Users/sergio/.rbenv/versions/3.2.2/bin/rspec:25:in `<main>'
     #
     #   Showing full backtrace because every line was filtered out.
     #   See docs for RSpec::Configuration#backtrace_exclusion_patterns and
     #   RSpec::Configuration#backtrace_inclusion_patterns for more information.
```

**Comment by pirj** [comment]
A wild guess. What if you remove this line?
expect(Mutex).to receive(:new)

**Comment by sergio-bobillier** [comment]
The result is the same with or without the expectation. I believe this condition arises from the setup in the `before` block.

**Comment by JonRowe** [comment]
This is different to the original bug and weirder. That was about the ability to stub mutex at all.
The call that causes the problem here is the stubbing of the synchronize method, if you remove:

```
allow(mocked_mutex).to receive(:synchronize).and_yield
```

The spec runs [or at least the version I set up in `rspec-support` does]

**Comment by sergio-bobillier** [comment]
@JonRowe Thanks, yes, that actually removes the issue, but the thing is that would also keep me from testing what I want.

The code I added was the minimum required to trigger the issue, however I'm interested in also testing what happens inside the `synchronize` block. In my tests I'm checking that some actions are performed under the "protection" of the Mutex. If I remove the `allow` then that code will never be executed.

It is not even possible to test that `synchronize` is being called over the `Mutex` instance, even without the `#and_yield` call the "Stack level too deep" error occurs.

Even if I remove the `allow` completely, any test checking the call to `synchronize` will trigger the same error, for example:

`spec/writer_spec.rb`:
```ruby
# frozen_string_literal: true

require_relative '../writer'

RSpec.describe Writer do
  subject(:writer) { described_class.new }

  let(:mocked_mutex) do
    instance_double(
      Mutex
    )
  end

  before do
    allow(Mutex).to receive(:new).and_return(mocked_mutex)
  end

  describe '#initialize' do
    it 'creates a new Mutex' do
      expect(Mutex).to receive(:new)
      writer
    end
  end

  describe '#write' do
    it 'uses the Mutex to synchronize the writing' do
      expect(mocked_mutex).to receive(:synchronize)
      writer.write("hello world!")
    end
  end
end
```

Now the test for `#initialize` passes, but the one for `#write` triggers the exact same issue.

And note that, just setting the expectation triggers the error, even if I remove the call to the actual method: `writer.write("hello world!")` the error will still be raised.

**Comment by JonRowe** [comment]
Apologies, I did not mean to imply you should remove the line, my intent was to highlight that it was the `synchronize` stub which was the root of the issue (to differnitate it from previous issues), as an aside your updated spec is essentially the same, its (likely) not a combination of allow and expect thats causing it.

**Comment by pirj** [comment]
We seem to be using `synchronize` in our code, and if it’s mocked to return a canned response - the behaviour is unpredictable.
Should we stash it for ourselves, too?

**Comment by JonRowe** [comment]
I don't think thats the issue, we stash `new` so that our instances of mutex shouldn't have that stub, and indeed I tried a version of where I did the same treatment to all methods with no changes. So either something is broken with how we stub / capture mutex on this or its a more complicated interaction, my quick investigation looked like it might be method visibility related

**Comment by nevinera** [comment]
I'm not sure I correctly understand how the 'stashing' is intended to operate - is [this](https://github.com/rspec/rspec-mocks/blob/main/lib/rspec/mocks/proxy.rb#L12-L15) the code that should be doing that? (If so, at this point, `Mutex` is already defined, so the `unless` skips that block - `Mutex` is always defined now).

I do find that the spec passes if I change this [line](https://github.com/rspec/rspec-mocks/blob/main/lib/rspec/mocks/proxy.rb#L30) to

```
@messages_received_mutex = RSpec::Support::Mutex.new
```
and would be happy to PR that (though I've no idea how to write a test that would exercise _that_)..

But if the problem is _actually_ that lines 13-14 needed to execute, what should we be checking? Updating [this line](https://github.com/rspec/rspec-mocks/blob/main/lib/rspec/mocks/proxy.rb#L12) to

```
unless defined?(Proxy::Mutex)
```

Also passes the spec, and appears to _perhaps_ reflect the intent - did the behavior of `defined?` change at some point recently?

Here's my stab, though I think that the spec might not be in the right place: https://github.com/rspec/rspec-mocks/pull/1575

**Comment by JonRowe** [comment]
> I'm not sure I correctly understand how the 'stashing' is intended to operate 
I mentioned this on the PR too but the code that I refer to as "stashing" is this https://github.com/rspec/rspec-support/blob/main/lib/rspec/support/reentrant_mutex.rb#L68

**Comment by JonRowe** [comment]
Fix released in 3.13.1
